src/foam/core/MultitonInfo.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.core;
+
+import java.util.Map;
+import java.util.HashMap;
+
+public class MultitonInfo<T>
+  implements Axiom, ContextFactory<T>
+{
+  Map<Object, T> instanceMap = new HashMap<Object, T>();
+
+  String name;
+  PropertyInfo p;
+
+  public MultitonInfo(String name, PropertyInfo p) {
+    this.name = name;
+    this.p = p;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public T getInstance(Map<String, Object> args, X x) {
```

Comment:
Isn't thread safe so could cause duplicate creates.

Reply:
Good eye. Made this method synchronized which should do the trick.